### PR TITLE
[dynamo][guards-cpp-refactor] NO_TENSOR_ALIASING guard

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120098
* #120096
* #120093
* #120123
* #120119
* #120091
* #120089
* #120068
* #120067
* __->__ #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1024,6 +1024,68 @@ class TENSOR_ALIASING : public RelationalGuard {
...
+ */
+class NO_TENSOR_ALIASING : public RelationalGuard {
+ public:
+  NO_TENSOR_ALIASING(
+      long unsigned int num_tensors,
+      py::object tensor_names,
+      py::object verbose_code_parts)
+      : RelationalGuard(verbose_code_parts),
+        _num_tensors(num_tensors),
+        _tensor_names(tensor_names) {}
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1038]: Call `unique_tensors.reseve(_num_tensors)`

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp]: Done.

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -1004,6 +1004,70 @@ class TENSOR_ALIASING : public RelationalGuard {
...
+    Py_INCREF(value);
+    auto insertion = _unique_tensors.insert({value, nullptr});
+    if (!insertion.second) {
+      // No need to clear _unique_tensors, reset_state will do
+      // it.
+      return false;
+    }
+    _counter++;
+    if (_counter == _num_tensors) {
+      reset_state();
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1038]: You should avoid the virtual method call here for slightly better performance.  I think you can use `final` for this.

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by anijain2305** [comment]
@jansel Ready for another review.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**